daemons/controld/controld_join_dc.c

```diff
@@ -673,23 +673,168 @@ finalize_sync_callback(xmlNode * msg, int call_id, int rc, xmlNode * output, voi
     }
 }
 
+/*!
+ * \internal
+ * \brief Free XML in a CIB client callback
+ *
+ * \param[in,out] xml  XML to free
+ */
 static void
-join_update_complete_callback(xmlNode * msg, int call_id, int rc, xmlNode * output, void *user_data)
+xml_callback_free_fn(void *xml)
 {
-    fsa_data_t *msg_data = NULL;
+    free_xml((xmlNode *) xml);
+}
 
-    if (rc == pcmk_ok) {
-        crm_debug("join-%d node history update (via CIB call %d) complete",
-                  current_join_id, call_id);
-        check_join_state(controld_globals.fsa_state, __func__);
+static void
+join_node_state_commit_callback(xmlNode *msg, int call_id, int rc,
+                                xmlNode *output, void *user_data)
+{
+    controld_forget_cib_replace_call(call_id);
 
-    } else {
-        crm_err("join-%d node history update (via CIB call %d) failed: %s "
+    if (rc != pcmk_ok) {
+        fsa_data_t *msg_data = NULL;    // for register_fsa_error() macro
+
+        crm_err("join-%d node history commit (via CIB call %d) failed: %s "
                 "(next transition may determine resource status incorrectly)",
```

Comment:
For now I kept the "(next transition...)" phrase in all of the error messages, just like we had already. I'm not sure that it's true though... if we hit an error in any of these callbacks, the controller exits. Presumably it should get the correct info when it restarts (and a different node will likely be DC).

Reply:
Yeah if we're exiting on error, no need for that -- but bump anywhere we're going to exit to crm_crit()